src/mouse_events.cpp

```diff
@@ -597,6 +597,48 @@ void mouse_handler::mouse_motion(int x, int y, const bool browse, bool update, m
 	}
 }
 
+// Hook for notifying lua game kernel of mouse button events. We pass button as
+// a serpaate argument than the original SDL event in order to manage touch
+// emulation (i.e., long touch = right click) and such.
+bool mouse_handler::mouse_button_event(const SDL_MouseButtonEvent& event, uint8_t button,
+									   map_location loc, bool click)
+{
+	if (gui().view_locked() || button < SDL_BUTTON_LEFT || button > SDL_BUTTON_X2) {
+		return false;
+	} else if (event.state > SDL_PRESSED || !gui().get_map().on_board(loc)) {
+		return false;
+	}
+
+	static const std::string buttons[6] = {
+		[0]					= "",
+		[SDL_BUTTON_LEFT]	= "left",
+		[SDL_BUTTON_MIDDLE]	= "middle",
+		[SDL_BUTTON_RIGHT]	= "right",
+		[SDL_BUTTON_X1]		= "x1",
+		[SDL_BUTTON_X2]		= "x2",
+	};
+	static const std::string events[3] = {
+		[SDL_RELEASED]		= "up",
+		[SDL_PRESSED]		= "down",
+		[2]					= "click",
+	};
+
+	if(game_lua_kernel* lk = pc_.gamestate().lua_kernel_.get()) {
+		lk->mouse_button_callback(loc, buttons[button], events[event.state]);
+
+		// Are we being asked to send a click event?
+		if (click) {
+			// Was both the up and down on the same map tile?
+			if (loc != drag_from_hex_) {
+				return false;
+			}
+			// We allow this event to be consumed, but not up/down
+			return lk->mouse_button_callback(loc, buttons[button], events[2]);
```

Comment:
Is `lk->mouse_button_callback`called twice when `click` is true?

Reply:
hmm okay i see

Reply:
So yes. You get one event for mouse down, one for mouse up and one for click, although not necessarily in that order. For left and middle buttons, you get down, click, then up. For the rest, you get down, up, then click.